server/image_pull.go

```diff
@@ -49,6 +83,17 @@ func (s *Server) PullImage(ctx context.Context, req *types.PullImageRequest) (*t
 		}
 	}
 
+	rhName, rh := s.GetRuntimeHandlerForPod(ctx, req.SandboxConfig)
+	if rh != nil {
+		if rh.RuntimePullImage {
+			// don't pull the image in crio, as it will be done by the runtime
```

Comment:
What happens if the image does not exist at all?

What happens if the network is flaky? This effectively blinds Kubelet’s visibility into pulls, and breaks its retry logic.

The credentials provided by Kubelet are discarded.

Reply:
Also, Kubelet can gather credentials for several sources and will _loop_ over them in `PullImage` until it finds credentials accepted by the registry.

I don’t particularly _like_ that behavior but that’s one more reason for `PullImage` to be supported properly, whatever that means (a RPC to the VM?).